### PR TITLE
Disallow some empty config URL settings

### PR DESCRIPTION
**Pull request by AndrewFerr**
If these settings are meant to be unspecified, then their entire parent
sections (`widgets` & `generic`) should be unspecified.

Signed-off-by: Andrew Ferrazzutti <andrewf@element.io>

### CODE REVIEW

File: src/Config/Config.ts [hunk 1]
```diff
@@ -323,7 +323,7 @@ export class BridgeWidgetConfig {
         if (yaml.disallowedIpRanges !== undefined && (!Array.isArray(yaml.disallowedIpRanges) || !yaml.disallowedIpRanges.every(s => typeof s === "string"))) {
             throw new ConfigError("widgets.disallowedIpRanges", "must be a string array");
         }
-        if (typeof yaml.publicUrl !== "string") {
+        if (typeof yaml.publicUrl !== "string" || !yaml.publicUrl) {
```
Comment (Half-Shot) [inline src/Config/Config.ts:326]: Minded to say we should go a step further and try to parse the URL with `new URL()`

File: src/Config/Config.ts [hunk 2]
```diff
@@ -267,7 +267,7 @@ export class BridgeConfigGenericWebhooks {
     public readonly waitForComplete?: boolean;
     public readonly enableHttpGet: boolean;
     constructor(yaml: BridgeGenericWebhooksConfigYAML) {
-        if (typeof yaml.urlPrefix !== "string") {
+        if (typeof yaml.urlPrefix !== "string" || !yaml.urlPrefix) {
```
Comment (Half-Shot) [inline src/Config/Config.ts:270]: Same as above.

**Comment by AndrewFerr** [comment]
The URLs are now parsed, and also stored as `URL`s (not just strings) in the config objects.

4bfc0f0 goes maybe a bit too far in allowing using `URL` config values for the dumped sample config file, and without needing to store a stringified copy of each URL.  Using dynamically assigned functions (instead of getters) was needed for the dumper to notice them & dump them in the proper order.

**Review by Half-Shot — COMMENTED** [review]
This looks fine, just the url concat behaviour needs looking at.

File: src/Connections/GenericHook.ts [hunk 3]
```diff
@@ -404,7 +404,7 @@ export class GenericHookConnection extends BaseConnection implements IConnection
                 name: this.state.name,
             },
             ...(showSecrets ? { secrets: {
-                url: new URL(this.hookId, this.config.urlPrefix),
+                url: new URL(this.hookId, this.config.parsedUrlPrefix),
```
[the remark this answers is not in the document]
Comment (AndrewFerr) [reply src/Connections/GenericHook.ts:407]: Don't worry, I did notice this (and bang my head about it :smile: )  **It works as long as the base URL has a trailing slash.**  That's why the `parsed` URL properties were all built by `makePrefixedUrl` (and is why `widgets.publicUrl` in the sample config gained a trailing slash).

[the remark this answers is not in the document]
Comment (Half-Shot) [reply src/Connections/GenericHook.ts:407]: Ah...I had never noticed that :D. Okay, ignore me :)

**Review by Half-Shot — APPROVED** [review]
I bow to your greater understanding of `URL` :)

**Pull request merged**